pygfx/utils/show.py

```diff
@@ -16,9 +17,243 @@
+        render is made.
+    after_render : Callable
+        A callback that will be executed during each draw call after a new
+        render is made.
+    draw_function : Callable
+        Replaces the draw callback with a custom one. If set both
+        `before_render` and `after_render` will have no effect.
+
+    """
+
+    before_render = RenderCallback()
+    after_render = RenderCallback()
+    draw_function = RenderCallback()
+
+    def __init__(
+        self,
+        canvas=None,
+        renderer=None,
+        controller=None,
+        camera=None,
+        before_render=None,
+        after_render=None,
+        draw_function=None,
+    ) -> None:
+        self.canvas = canvas
+        self.renderer = renderer
+        self.controller = controller
+        self.camera = camera
+        self.draw_function = draw_function or self.default_draw
+        self.scene = None
+
+        if before_render is not None:
+            self.before_render = before_render
+
+        if after_render is not None:
+            self.after_render = after_render
+
+    def default_draw(self):
+        if self.before_render is not None:
+            self.before_render()
+
+        self.controller.update_camera(self.camera)
+        self.renderer.render(self.scene, self.camera)
+
+        if self.after_render is not None:
+            self.after_render()
+
+        self.renderer.request_draw()
+
+    def show(
+        self,
+        object: WorldObject,
+        up=None,
+    ):
+        """Display a WorldObject
+
+        This function provides you with the basic scaffolding to visualize a given
+        WorldObject in a new window. While it does add scaffolding, it aims to be
+        fully customizable so that you can replace each piece as needed.
+
+        Parameters
+        ----------
+        object : gfx.WorldObject
+            The object to show. If it is not a :class:`gfx.Scene <pygfx.Scene>`
+            then Display will wrap it into a new scene containing lights and a
+            background.
+        up : gfx.Vector3
+            If set, and ``object`` does not contain a controller, set the camera
+            controller's up vector to this value.
+
+        Notes
+        -----
+        If you want to display multiple objects prefer :class:`gfx.Group
+        <pygfx.Group>` over :class:`gfx.Scene <pygfx.Scene>` to avoid the pitfall of
+        forgetting to add lights to the scene.
+
+        """
+
+        if isinstance(object, Scene):
+            scene = object
+        else:
+            scene = Scene()
+            scene.add(object)
+
+            dark_gray = np.array((169, 167, 168, 255)) / 255
+            light_gray = np.array((100, 100, 100, 255)) / 255
+
+            background = Background(None, BackgroundMaterial(light_gray, dark_gray))
+            scene.add(background)
+            scene.add(AmbientLight(), DirectionalLight())
+        self.scene = scene
+
+        if len(self.find_children(Light)) == 0:
```

Comment:
Could you use `any()` here? That would enable the more efficient generator implementation you hinted at

Reply:
hm, as long as `bool(gfx.Light(...)) == True` this should be possible. Its a bit weird to think of the boolean-ness of a WorldObject, but I like how `if any(self.find_children(Light))` looks syntax-wise 🤩 I will change it.